labscript/labscript.py

```diff
@@ -2173,8 +2179,90 @@ def generate_connection_table(hdf5_file):
     else:
         master_pseudoclock_name = compiler.master_pseudoclock.name
     dataset.attrs['master_pseudoclock'] = master_pseudoclock_name
-  
-  
+
+# Create a dictionary for caching results from vcs commands. The keys will be
+# the paths to files that are saved during save_labscripts(). The values will be
+# a list of tuples of the form (command, info, err); see the "Returns" section
+# of the _run_vcs_commands() docstring for more info. Also create a FileWatcher
+# instance for tracking when vcs results need updating. The callback will
+# replace the outdated cache entry with a new list of updated vcs commands and
+# outputs.
+_vcs_cache = {}
+_vcs_cache_rlock = threading.RLock()
+def _file_watcher_callback(name, info, event):
+    with _vcs_cache_rlock:
+        _vcs_cache[name] = _run_vcs_commands(name)
```

Comment:
So the lock is now being held for however long it takes to query the subprocess information. You can reduce lock contention by doing something like `temp = _run_vcs_commands(name)` outside of the lock and then only holding the lock when you place that into the `_vcs_cache` dictionary.

Reply:
Believe it or not I actually originally set things up as you're suggesting in both of these comments with `Lock` in place of `RLock`, but then changed it. For the lock here in `_file_watcher_callback()`, if the `_file_watcher` thread called `_file_watcher_callback()` then at least some of the cache data is out of date, so it is actually better for it to block the `save_labscripts()` thread from writing cached results to drive until the cache is updated.